File: TripKit/Classes/core/TransportKit/TKVehicular.m
//
//  TKVehicular.m
//  TripKit
//
//  Created by Adrian Schoenig on 19/03/2014.
//
//

#import "TKVehicular.h"

#import "TKTripKit.h"
#import "TKTransportKit.h"

@implementation TKVehicularHelper

+ (TKVehicleType)vehicleTypeForString:(NSString *)string
{
  if ([string isEqualToString:[self stringForVehicleType:TKVehicleType_Bicycle]])
    return TKVehicleType_Bicycle;
  if ([string isEqualToString:[self stringForVehicleType:TKVehicleType_Car]])
    return TKVehicleType_Car;
  if ([string isEqualToString:[self stringForVehicleType:TKVehicleType_Motorbike]])
    return TKVehicleType_Motorbike;
  if ([string isEqualToString:[self stringForVehicleType:TKVehicleType_SUV]])
    return TKVehicleType_SUV;
  return TKVehicleType_None;
}

+ (NSString *)stringForVehicleType:(TKVehicleType)vehicleType
{
  switch (vehicleType) {
    case TKVehicleType_Bicycle:
      return NSLocalizedStringFromTableInBundle(@"Bicycle", @"Shared", [TKTripKit bundle], "Text for vehicle of type: Bicycle");
      
    case TKVehicleType_Car:
      return NSLocalizedStringFromTableInBundle(@"Car", @"Shared", [TKTripKit bundle], "Text for vehicle of type: Car");
      
    case TKVehicleType_Motorbike:
      return NSLocalizedStringFromTableInBundle(@"Motorbike", @"Shared", [TKTripKit bundle], "Text for vehicle of type: Motorbike");
      
    case TKVehicleType_SUV:
      return NSLocalizedStringFromTableInBundle(@"SUV", @"Shared", [TKTripKit bundle], "Text for vehicle of type: SUV");

    case TKVehicleType_None:
      return nil;
  }
}

+ (TKImage *)iconForVehicle:(id<TKVehicular>)vehicle
{
  if (! [vehicle respondsToSelector:@selector(garage)]
      || ! [vehicle garage]) {
    return [TKStyleManager imageNamed:@"icon-mode-car-pool"];
  }
  
  TKVehicleType vehicleType = [vehicle vehicleType];
  switch (vehicleType) {
    case TKVehicleType_Bicycle:
      return [TKStyleManager imageNamed:@"icon-mode-bicycle"];
      
    case TKVehicleType_Car:
    case TKVehicleType_SUV:
      return [TKStyleManager imageNamed:@"icon-mode-car"];
      
    case TKVehicleType_Motorbike:
      return [TKStyleManager imageNamed:@"icon-mode-motorbike"];
      
    case TKVehicleType_None:
      return nil;
  }
}

+ (NSString *)titleForVehicle:(id<TKVehicular>)vehicle
{
  if (! [vehicle respondsToSelector:@selector(garage)]
      || ! [vehicle garage]) {
    return NSLocalizedStringFromTableInBundle(@"Getting a lift", @"Shared", [TKStyleManager bundle], nil);
  }
  
  NSString *name = [vehicle name];
  if (name.length > 0) {
    return name;
  } else {
    return [self stringForVehicleType:vehicle.vehicleType];
  }
}

@end

File: Sources/TripKit/model/TKVehicular.swift
//
//  TKVehicular.swift
//  TripKit
//
//  Created by Adrian Schönig on 17/8/21.
//  Copyright © 2021 SkedGo Pty Ltd. All rights reserved.
//

import Foundation

public enum TKVehicleType: Int {
  
  case unknown = 0
  case bicycle = 1
  case car = 2
  case motorbike = 3
  case SUV = 4
  case kickscooter = 5
  
}

public protocol TKVehicular {
  
  /// Optional name to use in the UI to refer to this vehicle.
  var name: String? { get }
   
  /// What kind of vehicle it is. Required field.
  var vehicleType: TKVehicleType { get }
  
  /// Where this vehicle is garaged. Can be `nil` but the algorithms won't try to
  /// take it back to the garage then.
  /// - note: `nil` is the same as getting a lift with someone
  var garage: MKAnnotation? { get }
  
  /// The unique identifier that identifies this vehicle.
  /// - note: Getting a lift instances don't have a UUID
  var vehicleID: UUID? { get }
}


extension TKVehicular {
  
  public var isCarPooling: Bool {
    garage == nil
  }
  
  public var icon: TKImage? {
    guard !isCarPooling else {
      return .iconModeCarPool
    }
    
    switch vehicleType {
    case .bicycle:      return .iconModeBicycle
    case .car,
         .SUV:          return .iconModeCar
    case .motorbike:    return .iconModeMotorbike
    case .kickscooter:  return .iconModeKickscooter
    case .unknown:      return nil
    }
  }
  
  public var title: String {
    if isCarPooling {
      return NSLocalizedString("Getting a lift", tableName: "Shared", bundle: .tripKit, comment: "Name for a segment or vehicle where you get a lift from someone.")
    }
    
    if let name = name, !name.isEmpty {
      return name
    } else {
      return vehicleType.title
    }
  }
  
}

extension TKVehicleType {
  public var title: String {
    switch self {
    case .bicycle: return Loc.VehicleTypeBicycle
    case .car: return Loc.VehicleTypeCar
    case .SUV: return Loc.VehicleTypeSUV
    case .motorbike: return Loc.VehicleTypeMotorbike
    case .kickscooter: return Loc.VehicleTypeKickScooter
    case .unknown: return Loc.Unknown
    }
  }
}
